### PR TITLE
Faster Angle initialisation.

### PR DESCRIPTION
**Pull request by mhvk**
In #3337, @embray proposed several changes to `SkyCoord` etc to speed up things. Unfortunately, that PR has been lingering, in part because its scope is relatively large, needing review from multiple people. But as he found that repeated initialization of `Angle` instances was a particular bottleneck, I thought I would address that at least. In this PR, I've optimized it such that for `Angle` or `Quantity` input, the creation is as fast as possible. (In reviewing the revised `__new__`, it may be easier to just look at the new version than at the diff...)

### CODE REVIEW

**Comment by taldcroft** [comment]
Do you have before/after timing information for various initialization options?

**Comment by mhvk** [comment]
It is meant particularly for passing through objects quickly that are already `Angle` (which happens a *lot* in `SkyCoord`). Currently:
```
q = np.arange(1000.) * u.deg
a = Angle(q)
%timeit Angle(q)
# 10000 loops, best of 3: 57.9 µs per loop
%timeit Angle(a)
# 10000 loops, best of 3: 58.8 µs per loop
```
These times change to 34.7 and 12.5 µs, resp.

**Comment by mhvk** [comment]
@taldcroft - I realised that even more relevantly than the above are processing time for `copy=False`. These also improve dramatically:

On current master (note, this is different machine, so cannot compare with above). 
```
In [7]: %timeit Angle(q, copy=False)
10000 loops, best of 3: 28.7 µs per loop

In [8]: %timeit Angle(a, copy=False)
10000 loops, best of 3: 29.1 µs per loop
```
With this PR: 13.7 and 1.62 µs.

I have not tested any other input, but it should be effectively unchanged, since for those whatever was in `__new__` did not dominate in the first place (or is unchanged).

File: astropy/coordinates/angles.py [hunk 1]
```diff
@@ -85,54 +85,43 @@ class Angle(u.Quantity):
...
-        if isinstance(angle, u.Quantity):
-            # This includes Angle subclasses as well
+        if not isinstance(angle, u.Quantity):
             if unit is not None:
-                angle = angle.to(unit).value
-            else:
-                unit = angle.unit
                 unit = cls._convert_unit_to_angle_unit(unit)
-                if not unit.is_equivalent(u.radian):
-                    raise u.UnitsError(
```
Comment (taldcroft) [inline astropy/coordinates/angles.py]: This must point to lack of coverage, but it seems some of the exception messages are changing.  I'm not too worried about this, but maybe this is a chance to fix up the coverage for exceptions here?

Comment (mhvk) [inline astropy/coordinates/angles.py]: The file is fully covered except for 6 lines for default cases/exceptions in `to_string`. I'd rather not go there in this PR... But I did add an explicit test for inputting something with a non-angular unit.

**Comment by taldcroft** [comment]
@mhvk - this looks good to me then.  I will confess that I have not stepped through the new logic in detail, but I believe that the existing testing is sufficient to prevent any serious regressions.

**Comment by taldcroft** [comment]
Except of course it needs a changelog.

**Comment by taldcroft** [comment]
@eteq @astrofrog - do you want time to review this PR or are you comfortable merging now?

**Comment by eteq** [comment]
Will review today.

**Comment by mhvk** [comment]
I'm happy to add a changelog entry, but don't think it really needs one, since this doesn't change the way the code behaves.

**Comment by taldcroft** [comment]
I definitely think that performance improvements should be noted!

**Comment by eteq** [comment]
Definitely agree with @taldcroft here that performance improvements should go in the changelog!

(Also, the possibility that the exceptions change is another reason to have *some* mention in the changelog, although no detail on that is necessary)

**Comment by mhvk** [comment]
OK, put a note under "other changes and additions".

File: astropy/coordinates/angles.py (hunk 1 above)
Comment (eteq) [inline astropy/coordinates/angles.py]: I think @taldcroft is suggesting you add tests for the *message*.  But I'm not sure I agree - in my head the type of the exception is part of the API, but the message itself is not (it's more like "documentation").  So I think it's fine as-is given that the type is checked.

**Comment by eteq** [comment]
This all looks good to me now.  Will go ahead and merge.  Thanks @mhvk!

**Pull request merged**